build_library/extract-initramfs-from-vmlinuz.sh

```diff
@@ -29,8 +29,12 @@ try_extract() {
     # cpio can do strange things when given garbage, so do a basic check.
     [[ $(head -c6 "$1") == 070701 ]] || return 0
 
-    # There may be multiple concatenated archives so try cpio till it fails.
-    while cpio --quiet --extract --make-directories --directory="${out}/rootfs-${ROOTFS_IDX}" --nonmatching 'dev/*' 2>/dev/null; do
+    while {
+        # cpio needs the directory to exist first. Fail if it's already there.
+        mkdir "${out}/rootfs-${ROOTFS_IDX}" || return $? &&
```

Comment:
I don't see the same behavior at all (checked cpio 2.15 and 2.13):
```
# strace -e mkdir,chdir -f cpio --quiet --extract --make-directories --directory="${out}/rootfs-${ROOTFS_IDX}" --nonmatching 'dev/*' <file.cpio
chdir("/out/rootfs-")                   = -1 ENOENT (No such file or directory)
mkdir("/out", 0755)                     = 0
mkdir("/out/rootfs-", 0755)             = 0
chdir("/out/rootfs-")                   = 0
```
So... something else going on here?

Reply:
It's not that it doesn't create the directory exactly. It does, but it somehow fails to chown/chmod it afterwards unless you create it first. I've been struggling to make sense of this, but it looks like it tries to reference the directory using the path given to `--directory` after changing into that directory.

With `2>/dev/null` removed:
```
sdk@flatcar-sdk ~ $ sudo rm -r initrd
sdk@flatcar-sdk ~ $ sudo ~/trunk/src/scripts/build_library/extract-initramfs-from-vmlinuz.sh ~/trunk/src/scripts/vmlinuz-6.6.88-flatcar initrd 
cpio: initrd/rootfs-0: Cannot change ownership to uid 0, gid 0: No such file or directory
cpio: initrd/rootfs-0: Cannot change mode to rwxr-xr-x: No such file or directory
done, found 1 rootfs(es)

sdk@flatcar-sdk ~ $ sudo rm -r initrd
sdk@flatcar-sdk ~ $ sudo mkdir -p initrd/rootfs-0
sdk@flatcar-sdk ~ $ sudo ~/trunk/src/scripts/build_library/extract-initramfs-from-vmlinuz.sh ~/trunk/src/scripts/vmlinuz-6.6.88-flatcar initrd 
cpio: initrd/rootfs-1: Cannot change ownership to uid 0, gid 0: No such file or directory
cpio: initrd/rootfs-1: Cannot change mode to rwxr-xr-x: No such file or directory
done, found 2 rootfs(es)

sdk@flatcar-sdk ~ $ sudo rm -r initrd
sdk@flatcar-sdk ~ $ sudo mkdir -p initrd/rootfs-{0,1}
sdk@flatcar-sdk ~ $ sudo ~/trunk/src/scripts/build_library/extract-initramfs-from-vmlinuz.sh ~/trunk/src/scripts/vmlinuz-6.6.88-flatcar initrd 
cpio: premature end of archive
done, found 2 rootfs(es)
```

With strace added:
```
sdk@flatcar-sdk ~ $ sudo rm -r initrd
sdk@flatcar-sdk ~ $ sudo ~/trunk/src/scripts/build_library/extract-initramfs-from-vmlinuz.sh ~/trunk/src/scripts/vmlinuz-6.6.88-flatcar initrd 2>&1 | fgrep -C5 rootfs-0
execve("/usr/sbin/cpio", ["cpio", "--quiet", "--extract", "--make-directories", "--directory=initrd/rootfs-0", "--nonmatching", "dev/*"], 0x7ffced3d8798 /* 21 vars */) = 0
brk(NULL)                               = 0x55f399f91000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7febdd82b000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=28370, ...}) = 0
--
mmap(NULL, 470080, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7febdd5b7000
close(3)                                = 0
geteuid()                               = 0
umask(000)                              = 022
fstat(0, {st_mode=S_IFREG|0644, st_size=102764032, ...}) = 0
chdir("initrd/rootfs-0")                = -1 ENOENT (No such file or directory)
getuid()                                = 0
newfstatat(AT_FDCWD, "initrd/rootfs-0", 0x7fff2cbdecf0, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "initrd", {st_mode=S_IFDIR|0755, st_size=6, ...}, 0) = 0
mkdir("initrd/rootfs-0", 0755)          = 0
newfstatat(AT_FDCWD, "initrd/rootfs-0", {st_mode=S_IFDIR|0755, st_size=6, ...}, 0) = 0
chdir("initrd/rootfs-0")                = 0
read(0, "070701001F11CE000041ED0000000000"..., 512) = 512
openat(AT_FDCWD, "/usr/lib64/gconv/gconv-modules.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=27012, ...}) = 0
mmap(NULL, 27012, PROT_READ, MAP_SHARED, 3, 0) = 0x7febdd824000
close(3)                                = 0
--
read(0, "\\\307\211m;y\315\237\r\354\2750\26\205Z\351F-M\263\356\273\323\33~9zV\2\200\237\217"..., 512) = 512
write(3, "dL\\\247\10\274S7\177\177[\316{9\262\10CC=\273\37\211\350\212\354\266\247\377C\353\357\256"..., 512) = 512
fchown(3, 0, 0)                         = 0
fchmod(3, 0100644)                      = 0
close(3)                                = 0
chown("initrd/rootfs-0", 0, 0)          = -1 ENOENT (No such file or directory)
write(2, "cpio: ", 6cpio: )                   = 6
write(2, "initrd/rootfs-0: Cannot change o"..., 56initrd/rootfs-0: Cannot change ownership to uid 0, gid 0) = 56
write(2, ": No such file or directory", 27: No such file or directory) = 27
write(2, "\n", 1
)                       = 1
chmod("initrd/rootfs-0", 040755)        = -1 ENOENT (No such file or directory)
write(2, "cpio: ", 6cpio: )                   = 6
write(2, "initrd/rootfs-0: Cannot change m"..., 48initrd/rootfs-0: Cannot change mode to rwxr-xr-x) = 48
write(2, ": No such file or directory", 27: No such file or directory) = 27
write(2, "\n", 1
)                       = 1
close(0)                                = 0
exit_group(2)
```

Reply:
Explicitly creating it first has the benefit of erroring out if it already exists. It would be confusing to extract into a directory with existing content if you thought it was previously empty.

Reply:
I'd like to get this in the next Alpha. This might be a bug in cpio, but I don't think we need to dwell on it. We were previously creating the directory anyway. I just removed that line when I rewrote the script because it initially seemed unnecessary.

Reply:
ok

Reply:
HEllo, I have a really hard time understanding this fix - seems to be the same fix I did https://github.com/flatcar/scripts/commit/0a8267f69b087b99d29b84193c8aa5c084030b45 here.

But the code has changed after my fix got merged and seemingly the same issue is back from those changes.
Now you are fixing it again?

Reply:
Note that my fix https://github.com/flatcar/scripts/commit/0a8267f69b087b99d29b84193c8aa5c084030b45 was required for the https://github.com/flatcar/scripts/pull/2300 (kernel upgrade to 6.12).

Reply:
I was aware of your fix, but it was necessary to restore multiple extractions after fixing the inclusion of the CPU microcode. I saw similarly odd results when I blindly tried to extract without checking for a cpio header, so I thought your issue would not occur if I did that. I'll check with your most recent 6.12 image now.

Reply:
It's okay, it's doing what it should now.

```
sdk@flatcar-sdk /tmp/initrd $ find rootfs-0/
rootfs-0/
rootfs-0/kernel
rootfs-0/kernel/x86
rootfs-0/kernel/x86/microcode
rootfs-0/kernel/x86/microcode/GenuineIntel.bin
rootfs-0/kernel/x86/microcode/AuthenticAMD.bin
rootfs-0/early_cpio
sdk@flatcar-sdk /tmp/initrd $ ls rootfs-1/
bin  dev  etc  init  lib  lib64  proc  root  run  sbin  shutdown  sys  sysroot  tmp  usr  var
```

To be clear, the CPU microcode is supposed to be in a separate cpio.

Reply:
I have only a cosmetic suggestion to avoid having `||` and `&&` next to each other without some grouping. Otherwise the scripts seems to be working for me for the latest nightly images (amd64 and arm64).

```suggestion
        { mkdir "${out}/rootfs-${ROOTFS_IDX}" || return $?; } &&
```